File: csharp/zipper/Zipper.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class BinTree
{
    public BinTree(int value, BinTree left, BinTree right)
        => (Value, Left, Right) = (value, left, right);

    public int Value { get; }
    public BinTree Left { get; }
    public BinTree Right { get; }

    public override bool Equals(object obj)
    {
        var that = obj as BinTree;
        if(that == null || that.Value != Value)
            return false;
        var leftMatch = (that.Left == null && Left == null) || (Left != null && Left.Equals(that.Left));
        var rightMatch = (that.Right == null && Right == null) || (Right != null && Right.Equals(that.Right));
        return leftMatch && rightMatch;
    }

    public override int GetHashCode() => base.GetHashCode();
}

public class Zipper
{
    private BinTree focus;
    private IEnumerable<(BinTree node, bool isLeft)> parents;

    public int Value() => focus.Value;
    public Zipper Left() => focus.Left != null ? FromTree(focus.Left, parents.Prepend((focus, true))) : null;
    public Zipper Right() => focus.Right != null ? FromTree(focus.Right, parents.Prepend((focus, false))) : null;
    public Zipper Up()
    {
        if(!parents.Any())
            return null;
        var parent = parents.First();
        var others = parents.Skip(1);

        var newFocus = new BinTree(parent.node.Value, 
            parent.isLeft ? focus : parent.node.Left, 
            parent.isLeft ? parent.node.Right : focus);
        return FromTree(newFocus, others);
    }

    public Zipper SetValue(int newValue) => FromTree(new BinTree(newValue, focus.Left, focus.Right), parents);

    public Zipper SetLeft(BinTree binTree) => FromTree(new BinTree(focus.Value, binTree, focus.Right), parents);

    public Zipper SetRight(BinTree binTree) => FromTree(new BinTree(focus.Value, focus.Left, binTree), parents);

    public BinTree ToTree()
    {
        var top = this;
        Zipper next = null;
        while((next = top.Up()) != null)
            top = next;
        return top.focus;
    }

    public static Zipper FromTree(BinTree tree, IEnumerable<(BinTree, bool)> parents = null) => 
        new Zipper { focus = tree, parents = parents ?? Enumerable.Empty<(BinTree, bool)>() };

    public override bool Equals(object obj) => (obj as Zipper).focus.Equals(focus);

    public override int GetHashCode() => base.GetHashCode();
}

File: fsharp/zipper/Zipper.fs
module Zipper

type Node = private { Value: int; Left: Node option; Right: Node option }
type Nav = private | Left | Right
type Zipper = private { Parents: (Node * Nav) list; Focus: Node }

let tree value left right = { Value = value; Left = left; Right = right }

let left zipper = 
    match zipper.Focus.Left with
    | None -> None
    | Some l -> Some { Focus = l; Parents = (zipper.Focus, Nav.Left)::zipper.Parents }
let right zipper = 
    match zipper.Focus.Right with
    | None -> None
    | Some r -> Some { Focus = r; Parents = (zipper.Focus, Nav.Right)::zipper.Parents }
let up zipper = 
    match zipper.Parents with 
    | [] -> None
    | h::tail -> 
        let newFocus = 
            if snd h = Nav.Left 
            then { fst h with Left = Some zipper.Focus }
            else { fst h with Right = Some zipper.Focus }
        Some { Focus = newFocus; Parents = tail }

let setValue value zipper = { zipper with Focus = { zipper.Focus with Value = value } }
let setLeft node zipper = { zipper with Focus = { zipper.Focus with Left = node } }
let setRight node zipper = { zipper with Focus = { zipper.Focus with Right = node } }

let fromTree tree = { Parents = []; Focus = tree }
let rec toTree zipper = 
    match up zipper with 
    | None -> zipper.Focus 
    | Some parent -> toTree parent
let value zipper = zipper.Focus.Value
